apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1065 @@
+    'nvidia-cusparselt==0.7.1; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia" in variant_namespaces',
+    'nvidia-nccl==2.27.3; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia" in variant_namespaces',
+
+    # CUDA 12.6
+    'nvidia-cuda-nvrtc==12.6.77; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.0" in variant_properties',
+    'nvidia-cuda-runtime==12.6.77; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.0" in variant_properties',
+
+    # CUDA 12.8
+    'nvidia-cuda-nvrtc==12.8.93; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.8" in variant_properties',
+    'nvidia-cuda-runtime==12.8.90; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.8" in variant_properties',
+
+    # CUDA 12.9
+    'nvidia-cuda-nvrtc==12.9.86; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.9" in variant_properties',
+    'nvidia-cuda-runtime==12.9.79; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.9" in variant_properties',
+]
+```
+
+</figure>
+
+There are three environment markers available for use: `variant_properties`
+corresponding to the set of the variant properties that the wheel was built for,
+`variant_features` combining them into enabled features (i.e. features that have
+at least one value), and `variant_namespaces` doing the same for namespaces.
+There has also been a discussion of adding a marker to match variant label
+specifically, i.e. `variant =​= "cu129"`, but so far there did not seem
+to emerge a really good use case for it, and it was deemed unnecessarily risky,
+as building the same variant with a different label would cause the dependency
+to stop applying.
+
+The discussion around environment markers surfaced more interesting problems.
+For a start, it was pointed out that there could be two different types
+of markers: markers matching the properties of the installed wheel (as outlined
+previously), and markers matching the current system state. For example,
+the latter could be used to install the CUDA plugin if CUDA is available without
+actually having to publish a separate CUDA variant of the main package (think
+JAX). However, this is a recent development and it has not been pursued yet.
+
+It was also pointed out that technically these kind of dependencies can create
+potential conflicts. For example, since technically a wheel can have multiple
+values for a property, you could end up pulling in two or three versions
+of `nvidia-cuda-runtime` simultaneously. This particularly causes problems
+when creating universal lockfiles — since they need to account for all valid
+combinations of environment markers.
```

Comment:
It feels like this needs a concluding sentence. Something like _"This is no different from how platform markers work however, and we concluded that it's up to package authors to ensure that there are no overlapping markers"._ Otherwise as a reader I'm taking away the impression that this design doesn't quite work or has an unsolved issue at least.

Reply:
Because it's not solved yet. The problem is exactly that platform markers are exclusive, whereas variant markers are not. I think "extras" are closer, so maybe that will work as a concluding sentence.

Reply:
Maybe conclude with something like "We are still working through some of the finer points of how universal resolving should work" then?

> The problem is exactly that platform markers are exclusive, whereas variant markers are not.

I am not convinced it's all that different; it's easy to cause the exact same problem with platform markers and have an incompatible set of dependency specifiers due to marker overlap. Individual platform markers are exclusive, but combinations of them aren't.